### ISSUE TITLE
Connection leak upon cancellation of reactive stream

### ISSUE THREAD

**Issue by GeorgiPetkov** [jOOQ/jOOQ#15462]
### Expected behavior

Whenever a cancel signal is received the following should happen (ordered by importance):
- the connection is released
- the cancellation is propagated (when used to wrap a publisher in a transaction)
- the query is canceled in the DB

### Actual behavior

Upon cancellation respectively:
- the connection may not be released when the cancellation is triggered fast enough, leading to a connection leak
- the cancellation is not propagated to wrapped publishers
- the query is not canceled in the DB

Furthermore, it should be possible to get into further problems if the connection is released and may be reused without actually canceling the old queries.

### Steps to reproduce the problem

I've provided tests focusing on the first and most important point regarding the connection leak (the rest could be addressed later, maybe with a separate issue).
All the tests execute `sleep(5 seconds)` and are parameterized to cancel the publisher in 0, 100, and 1000 millis.
The tests are 4, covering the combinations with/without using a connection pool and with/without the included workaround. The workaround never closes the connection on cancel (nor does it address the rest of the bullet points), but it's not leaking the connection (leaving it to be closed upon completion, in practice could be combined with `statement_timeout` for the time being).
The results are the same regardless of using a connection pool - if canceling very soon (less than ~30 millis) the connection is leaked, otherwise not. The workaround passes the tests (aside from the fact that the connection is closed/released after the query completes).

```java
import io.r2dbc.pool.ConnectionPool;
import io.r2dbc.pool.ConnectionPoolConfiguration;
import io.r2dbc.spi.Connection;
import io.r2dbc.spi.ConnectionFactories;
import io.r2dbc.spi.ConnectionFactory;
import io.r2dbc.spi.ConnectionFactoryOptions;
import org.jooq.DSLContext;
import org.jooq.Record1;
import org.jooq.SelectSelectStep;
import org.jooq.impl.DSL;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.params.ParameterizedTest;
import org.junit.jupiter.params.provider.MethodSource;
import org.mockito.Mockito;
import org.reactivestreams.Publisher;
import org.testcontainers.containers.PostgreSQLContainer;
import org.testcontainers.junit.jupiter.Container;
import org.testcontainers.junit.jupiter.Testcontainers;
import reactor.core.publisher.Flux;
import reactor.core.publisher.Mono;
import reactor.core.scheduler.Schedulers;
import reactor.test.StepVerifier;

import java.time.Duration;
import java.util.List;
import java.util.concurrent.atomic.AtomicReference;

import static org.jooq.impl.DSL.function;
import static org.jooq.impl.DSL.value;
import static org.jooq.impl.SQLDataType.OTHER;
import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.mockito.Mockito.*;
import static org.testcontainers.utility.DockerImageName.parse;

@Testcontainers
class ConnectionLeakingTest {

    private static final int SLEEP_SECONDS = 5;

    @Container
    private final PostgreSQLContainer<?> CONTAINER = new PostgreSQLContainer<>(parse("postgres:14.6-alpine"));

    private ConnectionFactory nonPooledConnectionFactory;
    private ConnectionPool pooledConnectionFactory;

    @BeforeEach
    void setup() {
        this.nonPooledConnectionFactory = ConnectionFactories.get(ConnectionFactoryOptions
                .parse(CONTAINER.getJdbcUrl().replaceFirst("^jdbc:", "r2dbc:"))
                .mutate()
                .option(ConnectionFactoryOptions.USER, CONTAINER.getUsername())
                .option(ConnectionFactoryOptions.PASSWORD, CONTAINER.getPassword())
                .build());
        this.pooledConnectionFactory = new ConnectionPool(ConnectionPoolConfiguration.builder(nonPooledConnectionFactory).build());
    }

    @ParameterizedTest
    @MethodSource("getCancellationDelays")
    void cancelShouldReleaseThePooledConnection(Duration cancellationDelay) throws InterruptedException {
        cancelQueryAfterDelay(sleepQuery(DSL.using(pooledConnectionFactory)), cancellationDelay);

        // give plenty of time for `acquire` and `release` to occur
        Thread.sleep(2000);
        assertConnectionsReleased();
    }

    @ParameterizedTest
    @MethodSource("getCancellationDelays")
    void workaroundShouldReleaseThePooledConnectionAfterCompletion(Duration cancellationDelay) throws InterruptedException {
        cancelQueryAfterDelay(notPropagatingCancellation(sleepQuery(DSL.using(pooledConnectionFactory))), cancellationDelay);

        // give plenty of time for `acquire` and `release` to occur
        // since we don't cancel the query we'll need more than the sleep duration
        Thread.sleep(SLEEP_SECONDS * 1000 + 2000);
        assertConnectionsReleased();
    }

    private void assertConnectionsReleased() {
        assertEquals(0, pooledConnectionFactory.getMetrics().get().acquiredSize());
    }

    @ParameterizedTest
    @MethodSource("getCancellationDelays")
    void cancelShouldCloseTheNonPooledConnection(Duration cancellationDelay) throws InterruptedException {
        ConnectionFactory connectionFactory = spy(nonPooledConnectionFactory);
        AtomicReference<Connection> connection = new AtomicReference<>();
        doAnswer(invocation -> Mono.from((Publisher<Connection>) invocation.callRealMethod()).map(Mockito::spy).doOnNext(connection::set))
                .when(connectionFactory).create();

        cancelQueryAfterDelay(sleepQuery(DSL.using(connectionFactory)), cancellationDelay);

        // give plenty of time for `acquire` and `release` to occur
        Thread.sleep(2000);
        assertConnectionClosed(connectionFactory, connection.get());
    }

    @ParameterizedTest
    @MethodSource("getCancellationDelays")
    void workaroundShouldCloseTheNonPooledConnectionAfterCompletion(Duration cancellationDelay) throws InterruptedException {
        ConnectionFactory connectionFactory = spy(nonPooledConnectionFactory);
        AtomicReference<Connection> connection = new AtomicReference<>();
        doAnswer(invocation -> Mono.from((Publisher<Connection>) invocation.callRealMethod()).map(Mockito::spy).doOnNext(connection::set))
                .when(connectionFactory).create();

        cancelQueryAfterDelay(notPropagatingCancellation(sleepQuery(DSL.using(connectionFactory))), cancellationDelay);

        // give plenty of time for `acquire` and `release` to occur
        // since we don't cancel the query we'll need more than the sleep duration
        Thread.sleep(SLEEP_SECONDS * 1000 + 2000);
        assertConnectionClosed(connectionFactory, connection.get());
    }

    private static void assertConnectionClosed(ConnectionFactory connectionFactory, Connection connection) {
        verify(connectionFactory).getMetadata();
        verify(connectionFactory).create();
        verify(connection).createStatement("select pg_sleep($1)");
        verify(connection).close();
        verifyNoMoreInteractions(connectionFactory, connection);
    }

    private static SelectSelectStep<Record1<Object>> sleepQuery(DSLContext dslContext) {
        return dslContext.select(function("pg_sleep", OTHER, value(SLEEP_SECONDS)));
    }

    private static void cancelQueryAfterDelay(Publisher<Record1<Object>> query, Duration cancellationDelay) {
        StepVerifier.create(query)
                .thenAwait(cancellationDelay)
                .thenCancel()
                .verify();
    }

    private static Iterable<Duration> getCancellationDelays() {
        return List.of(Duration.ZERO, Duration.ofMillis(100), Duration.ofMillis(1000));
    }

    // the workaround
    static <T> Flux<T> notPropagatingCancellation(Publisher<T> query) {
        return Flux.create(sink -> Flux.from(query)
                .doOnNext(sink::next)
                .doOnError(sink::error)
                .doOnComplete(sink::complete)
                .subscribeOn(Schedulers.parallel())
                .subscribe());
    }
}
```

### jOOQ Version

3.18.4, 3.18.5 (current latest)

### Database product and version

PostgreSQL 14.6

### Java Version

Java 17

### OS Version

Any

### JDBC driver name and version (include name if unofficial driver)

42.6.0

**Comment by petr-titov** [comment]
We have the same problem. We have some publishers created by jOOQ and Mono.from(). When we try to call Mono.zip with the publishers and one of them is failed, connections used by other publishers are leaked.

**Comment by SimoneGiusso** [comment]
@GeorgiPetkov I was interested in your tests because I may have the same issue. I don't know if you are familiar with Spring Boot, but if you inject in your test the `ConnectionPool` bean instantiated by Spring autoconfigure and assign it to `pooledConnectionFactory` your test `cancelShouldReleaseThePooledConnection` passes!

I may have the same problem described by @petr-titov.

**Comment by GeorgiPetkov** [comment]
@SimoneGiusso I am familiar with Spring Boot, but this is irrelevant. The point of the test is to be straight to the point with least dependencies possible. Regardless, I'm pretty sure it doesn't work with Spring Boot either because I had many troubles on production to get to this point and to pin-point the exact problem (and there's nothing special in how Spring Boot configures it).
It's about timing and you probably alterted a lot of things in the test, but the problem is still there even if somehow it isn't shown by your version of it with the same timing parameters. Sometimes just having more dependencies slows down at least the first test that uses a given library which can make the difference between failing and passing.

**Comment by SimoneGiusso** [comment]
@GeorgiPetkov can you try to put `:pool` in the url? 

Like [this](https://github.com/r2dbc/r2dbc-pool?tab=readme-ov-file#getting-started) and possibly have a runtime dependency on `r2dbc-pool`?

**Comment by GeorgiPetkov** [comment]
@SimoneGiusso Why do you want me to do that? You have the whole example. Can't you reproduce the problem? What's the point in me doing this?

**Comment by SimoneGiusso** [comment]
@GeorgiPetkov 

This configuration solve your issue with the pool:

```java
        ConnectionFactory pooledConnectionFactory = ConnectionFactories.get(ConnectionFactoryOptions.builder()
            .option(DRIVER, "pool")
            .option(PROTOCOL, "postgresql") // driver identifier, PROTOCOL is delegated as DRIVER by the pool.
            .option(HOST, CONTAINER.getHost())
            .option(PORT, 5432)
            .option(USER, CONTAINER.getUsername())
            .option(PASSWORD, CONTAINER.getPassword())
            .option(DATABASE, CONTAINER.getDatabaseName())
            .build());
```

For the non pooled tests, you are testing that `io.r2dbc.spi.Connection.close()` is called but it is stated in the [documentation](https://www.jooq.org/doc/latest/manual/sql-building/dsl-context/connection-vs-datasource/#reactive-querying) that jooq doesn't do it.

**Comment by GeorgiPetkov** [comment]
@SimoneGiusso 
1. I don't understand what exactly you're changing with the `DRIVER` and `PROTOCOL` options. Could you elaborate on how is this fixing the issue? What is the actual issue?
2. I'm still working with a connection factory in the non-pooled case. The documentation states that if someone else creates the connection and passes it to JooQ then JooQ won't close it because it's the responsibility of the creator to close it. But this is not the case - I'm using a `ConnectionFactory` that just happens to return a spy so I can observe what happened (or what didn't happen).

**Comment by SimoneGiusso** [comment]
> 
> 1. I don't understand what exactly you're changing with the `DRIVER` and `PROTOCOL` options. Could you elaborate on how is this fixing the issue? What is the actual issue?
> 2. I'm still working with a connection factory in the non-pooled case. The documentation states that if someone else creates the connection and passes it to JooQ then JooQ won't close it because it's the responsibility of the creator to close it. But this is not the case - I'm using a `ConnectionFactory` that just happens to return a spy so I can observe what happened (or what didn't happen).

https://github.com/reactor/reactor-pool/issues/136#issuecomment-834263342

**Comment by petr-titov** [comment]
I've created the [test](https://github.com/petr-titov/jooq-r2dbc-test/blob/main/src/test/java/com/example/JooqR2dbcTest.java) that highlight the problem: connections from the pool are licked if an error occurs (and the 'cancel' signal happens) in parallel executions.

**Comment by SimoneGiusso** [comment]
> connections from the pool are licked if an error occurs (and the 'cancel' signal happens) in parallel executions.

That's it basically what has been tested [here](https://github.com/r2dbc/r2dbc-pool/issues/198#issue-1885269011). There is no dependency on jooq.

**Comment by petr-titov** [comment]
I believe the problem is fixed in jOOQ v3.20.0

**Comment by petr-titov** [comment]
if to be more precise, v3.19.13 has fixed the problem

**Comment by lukaseder** [comment]
Thanks for your feedback, @petr-titov. So, this is likely a duplicate of this issue, then:

- https://github.com/jOOQ/jOOQ/issues/17094

**Comment by GeorgiPetkov** [comment]
Unfortunately, the issue is not resolved (probably not a duplicate of the fixed issue or there's more to it). I've tried on version 3.19.23 (latest used by latest Spring Boot) and 3.20.26 (currently latest) and the test from my initial description still fails as before.

@lukaseder  Please, reopen the issue (I have no permission to do so).

**Comment by lukaseder** [comment]
After setting up our MCVE template (https://github.com/jOOQ/jOOQ-mcve, it really helps to speed up reproducing things!), as follows, I'll be trying to reproduce things

```patch
diff --git a/jOOQ-mcve-java-postgres/pom.xml b/jOOQ-mcve-java-postgres/pom.xml
index fa0f0a5..5ab1ce5 100644
--- a/jOOQ-mcve-java-postgres/pom.xml
+++ b/jOOQ-mcve-java-postgres/pom.xml
@@ -58,10 +58,49 @@

         <!-- Testing -->
         <dependency>
-            <groupId>junit</groupId>
-            <artifactId>junit</artifactId>
-            <version>4.13.2</version>
-            <scope>test</scope>
+            <groupId>org.junit.jupiter</groupId>
+            <artifactId>junit-jupiter-api</artifactId>
+            <version>5.13.4</version>
+        </dependency>
+        <dependency>
+            <groupId>org.junit.jupiter</groupId>
+            <artifactId>junit-jupiter-params</artifactId>
+            <version>5.11.4</version>
+        </dependency>
+        <dependency>
+            <groupId>io.projectreactor</groupId>
+            <artifactId>reactor-core</artifactId>
+            <version>3.7.4</version>
+        </dependency>
+        <dependency>
+            <groupId>io.projectreactor</groupId>
+            <artifactId>reactor-test</artifactId>
+            <version>3.7.4</version>
+        </dependency>
+        <dependency>
+            <groupId>io.r2dbc</groupId>
+            <artifactId>r2dbc-pool</artifactId>
+            <version>1.0.2.RELEASE</version>
+        </dependency>
+        <dependency>
+            <groupId>org.postgresql</groupId>
+            <artifactId>r2dbc-postgresql</artifactId>
+            <version>1.0.7.RELEASE</version>
+        </dependency>
+        <dependency>
+            <groupId>org.mockito</groupId>
+            <artifactId>mockito-core</artifactId>
+            <version>5.14.2</version>
+        </dependency>
+        <dependency>
+            <groupId>org.testcontainers</groupId>
+            <artifactId>testcontainers</artifactId>
+            <version>1.21.3</version>
+        </dependency>
+        <dependency>
+            <groupId>org.testcontainers</groupId>
+            <artifactId>junit-jupiter</artifactId>
+            <version>1.20.4</version>
         </dependency>
     </dependencies>

diff --git a/jOOQ-mcve-java-postgres/src/test/java/org/jooq/mcve/test/java/postgres/ConnectionLeakingTest.java b/jOOQ-mcve-java-postgres/src/test/java/org/jooq/mcve/test/java/postgres/ConnectionLeakingTest.java
index b2726eb..cc918ea 100644
--- a/jOOQ-mcve-java-postgres/src/test/java/org/jooq/mcve/test/java/postgres/ConnectionLeakingTest.java
+++ b/jOOQ-mcve-java-postgres/src/test/java/org/jooq/mcve/test/java/postgres/ConnectionLeakingTest.java
@@ -1,4 +1,147 @@
 package org.jooq.mcve.test.java.postgres;

+import io.r2dbc.pool.ConnectionPool;
+import io.r2dbc.pool.ConnectionPoolConfiguration;
+import io.r2dbc.spi.Connection;
+import io.r2dbc.spi.ConnectionFactories;
+import io.r2dbc.spi.ConnectionFactory;
+import io.r2dbc.spi.ConnectionFactoryOptions;
+import org.jooq.DSLContext;
+import org.jooq.Record1;
+import org.jooq.SelectSelectStep;
+import org.jooq.impl.DSL;
+import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.params.ParameterizedTest;
+import org.junit.jupiter.params.provider.MethodSource;
+import org.mockito.Mockito;
+import org.reactivestreams.Publisher;
+import org.testcontainers.containers.PostgreSQLContainer;
+import org.testcontainers.junit.jupiter.Container;
+import org.testcontainers.junit.jupiter.Testcontainers;
+import reactor.core.publisher.Flux;
+import reactor.core.publisher.Mono;
+import reactor.core.scheduler.Schedulers;
+import reactor.test.StepVerifier;
+
+import java.time.Duration;
+import java.util.List;
+import java.util.concurrent.atomic.AtomicReference;
+
+import static org.jooq.impl.DSL.function;
+import static org.jooq.impl.DSL.value;
+import static org.jooq.impl.SQLDataType.OTHER;
+import static org.junit.jupiter.api.Assertions.assertEquals;
+import static org.mockito.Mockito.*;
+import static org.testcontainers.utility.DockerImageName.parse;
+
+@Testcontainers
 public class ConnectionLeakingTest {
-}
+
+    private static final int SLEEP_SECONDS = 5;
+
+    @Container
+    private final PostgreSQLContainer<?> CONTAINER = new PostgreSQLContainer<>(parse("postgres:14.6-alpine"));
+
+    private ConnectionFactory nonPooledConnectionFactory;
+    private ConnectionPool pooledConnectionFactory;
+
+    @BeforeEach
+    void setup() {
+        this.nonPooledConnectionFactory = ConnectionFactories.get(ConnectionFactoryOptions
+                .parse(CONTAINER.getJdbcUrl().replaceFirst("^jdbc:", "r2dbc:"))
+                .mutate()
+                .option(ConnectionFactoryOptions.USER, CONTAINER.getUsername())
+                .option(ConnectionFactoryOptions.PASSWORD, CONTAINER.getPassword())
+                .build());
+        this.pooledConnectionFactory = new ConnectionPool(ConnectionPoolConfiguration.builder(nonPooledConnectionFactory).build());
+    }
+
+    @ParameterizedTest
+    @MethodSource("getCancellationDelays")
+    void cancelShouldReleaseThePooledConnection(Duration cancellationDelay) throws InterruptedException {
+        cancelQueryAfterDelay(sleepQuery(DSL.using(pooledConnectionFactory)), cancellationDelay);
+
+        // give plenty of time for `acquire` and `release` to occur
+        Thread.sleep(2000);
+        assertConnectionsReleased();
+    }
+
+    @ParameterizedTest
+    @MethodSource("getCancellationDelays")
+    void workaroundShouldReleaseThePooledConnectionAfterCompletion(Duration cancellationDelay) throws InterruptedException {
+        cancelQueryAfterDelay(notPropagatingCancellation(sleepQuery(DSL.using(pooledConnectionFactory))), cancellationDelay);
+
+        // give plenty of time for `acquire` and `release` to occur
+        // since we don't cancel the query we'll need more than the sleep duration
+        Thread.sleep(SLEEP_SECONDS * 1000 + 2000);
+        assertConnectionsReleased();
+    }
+
+    private void assertConnectionsReleased() {
+        assertEquals(0, pooledConnectionFactory.getMetrics().get().acquiredSize());
+    }
+
+    @ParameterizedTest
+    @MethodSource("getCancellationDelays")
+    void cancelShouldCloseTheNonPooledConnection(Duration cancellationDelay) throws InterruptedException {
+        ConnectionFactory connectionFactory = spy(nonPooledConnectionFactory);
+        AtomicReference<Connection> connection = new AtomicReference<>();
+        doAnswer(invocation -> Mono.from((Publisher<Connection>) invocation.callRealMethod()).map(Mockito::spy).doOnNext(connection::set))
+                .when(connectionFactory).create();
+
+        cancelQueryAfterDelay(sleepQuery(DSL.using(connectionFactory)), cancellationDelay);
+
+        // give plenty of time for `acquire` and `release` to occur
+        Thread.sleep(2000);
+        assertConnectionClosed(connectionFactory, connection.get());
+    }
+
+    @ParameterizedTest
+    @MethodSource("getCancellationDelays")
+    void workaroundShouldCloseTheNonPooledConnectionAfterCompletion(Duration cancellationDelay) throws InterruptedException {
+        ConnectionFactory connectionFactory = spy(nonPooledConnectionFactory);
+        AtomicReference<Connection> connection = new AtomicReference<>();
+        doAnswer(invocation -> Mono.from((Publisher<Connection>) invocation.callRealMethod()).map(Mockito::spy).doOnNext(connection::set))
+                .when(connectionFactory).create();
+
+        cancelQueryAfterDelay(notPropagatingCancellation(sleepQuery(DSL.using(connectionFactory))), cancellationDelay);
+
+        // give plenty of time for `acquire` and `release` to occur
+        // since we don't cancel the query we'll need more than the sleep duration
+        Thread.sleep(SLEEP_SECONDS * 1000 + 2000);
+        assertConnectionClosed(connectionFactory, connection.get());
+    }
+
+    private static void assertConnectionClosed(ConnectionFactory connectionFactory, Connection connection) {
+        verify(connectionFactory).getMetadata();
+        verify(connectionFactory).create();
+        verify(connection).createStatement("select pg_sleep($1)");
+        verify(connection).close();
+        verifyNoMoreInteractions(connectionFactory, connection);
+    }
+
+    private static SelectSelectStep<Record1<Object>> sleepQuery(DSLContext dslContext) {
+        return dslContext.select(function("pg_sleep", OTHER, value(SLEEP_SECONDS)));
+    }
+
+    private static void cancelQueryAfterDelay(Publisher<Record1<Object>> query, Duration cancellationDelay) {
+        StepVerifier.create(query)
+                .thenAwait(cancellationDelay)
+                .thenCancel()
+                .verify();
+    }
+
+    private static Iterable<Duration> getCancellationDelays() {
+        return List.of(Duration.ZERO, Duration.ofMillis(100), Duration.ofMillis(1000));
+    }
+
+    // the workaround
+    static <T> Flux<T> notPropagatingCancellation(Publisher<T> query) {
+        return Flux.create(sink -> Flux.from(query)
+                .doOnNext(sink::next)
+                .doOnError(sink::error)
+                .doOnComplete(sink::complete)
+                .subscribeOn(Schedulers.parallel())
+                .subscribe());
+    }
+}
```

There's a single test that fails with this error:

```
org.mockito.exceptions.misusing.NullInsteadOfMockException: 
Argument passed to verify() should be a mock but is null!
Examples of correct verifications:
    verify(mock).someMethod();
    verify(mock, times(10)).someMethod();
    verify(mock, atLeastOnce()).someMethod();
    not: verify(mock.someMethod());
Also, if you use @Mock annotation don't miss openMocks()

	at org.jooq.mcve.test.java.postgres.ConnectionLeakingTest.assertConnectionClosed(ConnectionLeakingTest.java:118)
	at org.jooq.mcve.test.java.postgres.ConnectionLeakingTest.cancelShouldCloseTheNonPooledConnection(ConnectionLeakingTest.java:96)
	at java.base/java.lang.reflect.Method.invoke(Method.java:580)
	at java.base/java.util.Optional.ifPresent(Optional.java:178)
	at java.base/java.util.stream.ForEachOps$ForEachOp$OfRef.accept(ForEachOps.java:184)
```

There are some exceptions in the logs, but hard to say if test errors or actual errors. I'll have to investigate.

**Comment by lukaseder** [comment]
I believe this would be a better test:

```java
    private static void assertConnectionClosed(ConnectionFactory connectionFactory, Connection connection) {
        verify(connectionFactory).getMetadata();
        verify(connectionFactory).create();

        if (connection != null) {
            verify(connection).createStatement("select pg_sleep($1)");
            verify(connection).close();
            verifyNoMoreInteractions(connectionFactory, connection);
        }
        else
            verifyNoMoreInteractions(connectionFactory);
    }
```

If the connection is never acquired (because the cancellation preceds acquisition), then we shouldn't assert its release.

On a different machine, I get this additional test failure for `cancellationDelay = PT0.1S`. I believe this is what this bug report is really about:

```
Wanted but not invoked:
postgresqlConnection.createStatement(
    "select pg_sleep($1)"
);
-> at io.r2dbc.postgresql.PostgresqlConnection.createStatement(PostgresqlConnection.java:258)
Actually, there were zero interactions with this mock.

Wanted but not invoked:
postgresqlConnection.createStatement(
    "select pg_sleep($1)"
);
-> at io.r2dbc.postgresql.PostgresqlConnection.createStatement(PostgresqlConnection.java:258)
Actually, there were zero interactions with this mock.

	at io.r2dbc.postgresql.PostgresqlConnection.createStatement(PostgresqlConnection.java:258)
	at io.r2dbc.postgresql.PostgresqlConnection.createStatement(PostgresqlConnection.java:67)
	at org.jooq.mcve.test.java.postgres.ConnectionLeakingTest.assertConnectionClosed(ConnectionLeakingTest.java:118)
	at org.jooq.mcve.test.java.postgres.ConnectionLeakingTest.cancelShouldCloseTheNonPooledConnection(ConnectionLeakingTest.java:96)
```

Will continue investigation.

**Comment by lukaseder** [comment]
I find the tests using mocks inconclusive. With R2DBC, everything is lazy. The call to `ConnectionFactory.create()` creates a `Publisher`, not a `Subscription`. So, even if `ConnectionFactory.create()` is called, you cannot assert that `Connection.createStatement()` is called as well.

The important check to do here is whether any `Publisher` have been subscribed to, in case of which a `Connection` needs to be closed, not otherwise.

I've updated your test to check for precisely this, and the checks don't fail. Look for `assertEquals(0L, openR2dbcConnectionCounter.get());` tests below:

```java
package org.jooq.mcve.test.java.postgres;

import io.r2dbc.pool.ConnectionPool;
import io.r2dbc.pool.ConnectionPoolConfiguration;
import io.r2dbc.spi.*;
import org.jooq.DSLContext;
import org.jooq.Record1;
import org.jooq.SelectSelectStep;
import org.jooq.impl.DSL;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.params.ParameterizedTest;
import org.junit.jupiter.params.provider.MethodSource;
import org.mockito.Mockito;
import org.reactivestreams.Publisher;
import org.reactivestreams.Subscriber;
import org.reactivestreams.Subscription;
import org.testcontainers.containers.PostgreSQLContainer;
import org.testcontainers.junit.jupiter.Container;
import org.testcontainers.junit.jupiter.Testcontainers;
import reactor.core.publisher.Flux;
import reactor.core.publisher.Mono;
import reactor.core.scheduler.Schedulers;
import reactor.test.StepVerifier;

import java.time.Duration;
import java.util.List;
import java.util.concurrent.atomic.AtomicLong;
import java.util.concurrent.atomic.AtomicReference;
import java.util.function.Consumer;

import static org.jooq.impl.DSL.function;
import static org.jooq.impl.DSL.value;
import static org.jooq.impl.SQLDataType.OTHER;
import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.mockito.Mockito.*;
import static org.testcontainers.utility.DockerImageName.parse;

@Testcontainers
public class ConnectionLeakingTest {

    private static final int SLEEP_SECONDS = 5;

    @Container
    private final PostgreSQLContainer<?> CONTAINER = new PostgreSQLContainer<>(parse("postgres:14.6-alpine"));

    private ConnectionFactory nonPooledConnectionFactory;
    private ConnectionPool pooledConnectionFactory;

    @BeforeEach
    void setup() {
        this.nonPooledConnectionFactory = ConnectionFactories.get(ConnectionFactoryOptions
                .parse(CONTAINER.getJdbcUrl().replaceFirst("^jdbc:", "r2dbc:"))
                .mutate()
                .option(ConnectionFactoryOptions.USER, CONTAINER.getUsername())
                .option(ConnectionFactoryOptions.PASSWORD, CONTAINER.getPassword())
                .build());
        this.pooledConnectionFactory = new ConnectionPool(ConnectionPoolConfiguration.builder(nonPooledConnectionFactory).build());
    }
    public static final <T> Subscriber<T> subscriber(
            Consumer<? super Subscription> subscription,
            Consumer<? super T> onNext,
            Consumer<? super Throwable> onError,
            Runnable onComplete
    ) {
        return new Subscriber<T>() {
            @Override
            public void onSubscribe(Subscription s) {
                subscription.accept(s);
            }

            @Override
            public void onNext(T t) {
                onNext.accept(t);
            }

            @Override
            public void onError(Throwable t) {
                onError.accept(t);
            }

            @Override
            public void onComplete() {
                onComplete.run();
            }
        };
    }

    public AtomicLong openR2dbcConnectionCounter = new AtomicLong();

    ConnectionFactory wrap(ConnectionFactory delegateCF) {
        return new ConnectionFactory() {
            @Override
            public Publisher<? extends io.r2dbc.spi.Connection> create() {
                return delegatePCS -> delegateCF.create().subscribe(subscriber(
                        delegatePCS::onSubscribe,
                        delegateC -> {
                            System.out.println("Open R2DBC connections (open) : " + openR2dbcConnectionCounter.incrementAndGet());
                            delegatePCS.onNext(new org.jooq.tools.r2dbc.LoggingConnection(new io.r2dbc.spi.Connection() {

                                // ---------------------------------------------------------------------
                                // 0.9.0.M1 API
                                // ---------------------------------------------------------------------

                                @Override
                                public Publisher<Void> beginTransaction() {
                                    return delegateC.beginTransaction();
                                }

                                @Override
                                public Publisher<Void> beginTransaction(TransactionDefinition definition) {
                                    return delegateC.beginTransaction(definition);
                                }

                                @Override
                                public Publisher<Void> close() {
                                    Publisher<Void> delegatePV = delegateC.close();
                                    return delegatePVS -> delegatePV.subscribe(subscriber(
                                            delegatePVS::onSubscribe,
                                            delegatePVS::onNext,
                                            t -> {
                                                System.out.println("Open R2DBC connections (error): " + openR2dbcConnectionCounter.get());
                                                t.printStackTrace();
                                                delegatePVS.onError(t);
                                            },
                                            () -> {
                                                System.out.println("Open R2DBC connections (close): " + openR2dbcConnectionCounter.decrementAndGet());
                                                delegatePVS.onComplete();
                                            }
                                    ));
                                }

                                @Override
                                public Publisher<Void> commitTransaction() {
                                    return delegateC.commitTransaction();
                                }

                                @Override
                                public Batch createBatch() {
                                    return delegateC.createBatch();
                                }

                                @Override
                                public Publisher<Void> createSavepoint(String name) {
                                    return delegateC.createSavepoint(name);
                                }

                                @Override
                                public io.r2dbc.spi.Statement createStatement(String sql) {
                                    return delegateC.createStatement(sql);
                                }

                                @Override
                                public boolean isAutoCommit() {
                                    return delegateC.isAutoCommit();
                                }

                                @Override
                                public ConnectionMetadata getMetadata() {
                                    return delegateC.getMetadata();
                                }

                                @Override
                                public IsolationLevel getTransactionIsolationLevel() {
                                    return delegateC.getTransactionIsolationLevel();
                                }

                                @Override
                                public Publisher<Void> releaseSavepoint(String name) {
                                    return delegateC.releaseSavepoint(name);
                                }

                                @Override
                                public Publisher<Void> rollbackTransaction() {
                                    return delegateC.rollbackTransaction();
                                }

                                @Override
                                public Publisher<Void> rollbackTransactionToSavepoint(String name) {
                                    return delegateC.rollbackTransactionToSavepoint(name);
                                }

                                @Override
                                public Publisher<Void> setAutoCommit(boolean autoCommit) {
                                    return delegateC.setAutoCommit(autoCommit);
                                }

                                @Override
                                public Publisher<Void> setTransactionIsolationLevel(IsolationLevel isolationLevel) {
                                    return delegateC.setTransactionIsolationLevel(isolationLevel);
                                }

                                @Override
                                public Publisher<Boolean> validate(ValidationDepth depth) {
                                    return delegateC.validate(depth);
                                }

                                // ---------------------------------------------------------------------
                                // 0.9.0.M2 API
                                // ---------------------------------------------------------------------

                                @Override
                                public Publisher<Void> setLockWaitTimeout(Duration timeout) {
                                    return delegateC.setLockWaitTimeout(timeout);
                                }

                                @Override
                                public Publisher<Void> setStatementTimeout(Duration timeout) {
                                    return delegateC.setStatementTimeout(timeout);
                                }
                            }));
                        },
                        delegatePCS::onError,
                        delegatePCS::onComplete
                ));
            }

            @Override
            public ConnectionFactoryMetadata getMetadata() {
                return delegateCF.getMetadata();
            }
        };
    }

    @ParameterizedTest
    @MethodSource("getCancellationDelays")
    void cancelShouldReleaseThePooledConnection(Duration cancellationDelay) throws InterruptedException {
        cancelQueryAfterDelay(sleepQuery(DSL.using(wrap(pooledConnectionFactory))), cancellationDelay);

        // give plenty of time for `acquire` and `release` to occur
        Thread.sleep(2000);
        assertEquals(0L, openR2dbcConnectionCounter.get());
        assertConnectionsReleased();
    }

    @ParameterizedTest
    @MethodSource("getCancellationDelays")
    void workaroundShouldReleaseThePooledConnectionAfterCompletion(Duration cancellationDelay) throws InterruptedException {
        cancelQueryAfterDelay(notPropagatingCancellation(sleepQuery(DSL.using(wrap(pooledConnectionFactory)))), cancellationDelay);

        // give plenty of time for `acquire` and `release` to occur
        // since we don't cancel the query we'll need more than the sleep duration
        Thread.sleep(SLEEP_SECONDS * 1000 + 2000);
        assertEquals(0L, openR2dbcConnectionCounter.get());
        assertConnectionsReleased();
    }

    private void assertConnectionsReleased() {
        assertEquals(0, pooledConnectionFactory.getMetrics().get().acquiredSize());
    }

    @ParameterizedTest
    @MethodSource("getCancellationDelays")
    void cancelShouldCloseTheNonPooledConnection(Duration cancellationDelay) throws InterruptedException {
        ConnectionFactory connectionFactory = spy(wrap(nonPooledConnectionFactory));
        AtomicReference<Connection> connection = new AtomicReference<>();
        doAnswer(invocation -> Mono.from((Publisher<Connection>) invocation.callRealMethod()).map(Mockito::spy).doOnNext(connection::set))
                .when(connectionFactory).create();

        cancelQueryAfterDelay(sleepQuery(DSL.using(connectionFactory)), cancellationDelay);

        // give plenty of time for `acquire` and `release` to occur
        Thread.sleep(2000);
        assertEquals(0L, openR2dbcConnectionCounter.get());
        assertConnectionClosed(connectionFactory, connection.get());
    }

    @ParameterizedTest
    @MethodSource("getCancellationDelays")
    void workaroundShouldCloseTheNonPooledConnectionAfterCompletion(Duration cancellationDelay) throws InterruptedException {
        ConnectionFactory connectionFactory = spy(wrap(nonPooledConnectionFactory));
        AtomicReference<Connection> connection = new AtomicReference<>();
        doAnswer(invocation -> Mono.from((Publisher<Connection>) invocation.callRealMethod()).map(Mockito::spy).doOnNext(connection::set))
                .when(connectionFactory).create();

        cancelQueryAfterDelay(notPropagatingCancellation(sleepQuery(DSL.using(connectionFactory))), cancellationDelay);

        // give plenty of time for `acquire` and `release` to occur
        // since we don't cancel the query we'll need more than the sleep duration
        Thread.sleep(SLEEP_SECONDS * 1000 + 2000);
        assertEquals(0L, openR2dbcConnectionCounter.get());
        assertConnectionClosed(connectionFactory, connection.get());
    }

    private static void assertConnectionClosed(ConnectionFactory connectionFactory, Connection connection) {
        verify(connectionFactory).getMetadata();
        verify(connectionFactory).create();

        if (connection != null) {
            verify(connection).createStatement("select pg_sleep($1)");
            verify(connection).close();
            verifyNoMoreInteractions(connectionFactory, connection);
        }
        else
            verifyNoMoreInteractions(connectionFactory);
    }

    private static SelectSelectStep<Record1<Object>> sleepQuery(DSLContext dslContext) {
        return dslContext.select(function("pg_sleep", OTHER, value(SLEEP_SECONDS)));
    }

    private static void cancelQueryAfterDelay(Publisher<Record1<Object>> query, Duration cancellationDelay) {
        StepVerifier.create(query)
                // .thenRequest(1)
                .thenAwait(cancellationDelay)
                .thenCancel()
                .verify();
    }

    private static Iterable<Duration> getCancellationDelays() {
        return List.of(Duration.ZERO, Duration.ofMillis(100), Duration.ofMillis(1000));
    }

    // the workaround
    static <T> Flux<T> notPropagatingCancellation(Publisher<T> query) {
        return Flux.create(sink -> Flux.from(query)
                .doOnNext(sink::next)
                .doOnError(sink::error)
                .doOnComplete(sink::complete)
                .subscribeOn(Schedulers.parallel())
                .subscribe());
    }
}
```

I'm not negating there might be a problem, but I would need to be able to reliably reproduce it. Your test does not do that.

**Comment by lukaseder** [comment]
Here's another version of the test case with mockito removed completely, as it is just a distraction:

```java
package org.jooq.mcve.test.java.postgres;

import io.r2dbc.pool.ConnectionPool;
import io.r2dbc.pool.ConnectionPoolConfiguration;
import io.r2dbc.spi.*;
import org.jooq.DSLContext;
import org.jooq.Record1;
import org.jooq.SelectSelectStep;
import org.jooq.impl.DSL;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.params.ParameterizedTest;
import org.junit.jupiter.params.provider.MethodSource;
import org.reactivestreams.Publisher;
import org.reactivestreams.Subscriber;
import org.reactivestreams.Subscription;
import org.testcontainers.containers.PostgreSQLContainer;
import org.testcontainers.junit.jupiter.Container;
import org.testcontainers.junit.jupiter.Testcontainers;
import reactor.core.publisher.Flux;
import reactor.core.scheduler.Schedulers;
import reactor.test.StepVerifier;

import java.time.Duration;
import java.util.List;
import java.util.concurrent.atomic.AtomicLong;
import java.util.function.Consumer;

import static org.jooq.impl.DSL.function;
import static org.jooq.impl.DSL.value;
import static org.jooq.impl.SQLDataType.OTHER;
import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.testcontainers.utility.DockerImageName.parse;

@Testcontainers
public class ConnectionLeakingTest {

    private static final int SLEEP_SECONDS = 5;

    @Container
    private final PostgreSQLContainer<?> CONTAINER = new PostgreSQLContainer<>(parse("postgres:14.6-alpine"));

    private ConnectionFactory nonPooledConnectionFactory;
    private ConnectionPool pooledConnectionFactory;

    @BeforeEach
    void setup() {
        this.nonPooledConnectionFactory = ConnectionFactories.get(ConnectionFactoryOptions
                .parse(CONTAINER.getJdbcUrl().replaceFirst("^jdbc:", "r2dbc:"))
                .mutate()
                .option(ConnectionFactoryOptions.USER, CONTAINER.getUsername())
                .option(ConnectionFactoryOptions.PASSWORD, CONTAINER.getPassword())
                .build());
        this.pooledConnectionFactory = new ConnectionPool(ConnectionPoolConfiguration.builder(nonPooledConnectionFactory).build());
    }
    public static final <T> Subscriber<T> subscriber(
            Consumer<? super Subscription> subscription,
            Consumer<? super T> onNext,
            Consumer<? super Throwable> onError,
            Runnable onComplete
    ) {
        return new Subscriber<T>() {
            @Override
            public void onSubscribe(Subscription s) {
                subscription.accept(s);
            }

            @Override
            public void onNext(T t) {
                onNext.accept(t);
            }

            @Override
            public void onError(Throwable t) {
                onError.accept(t);
            }

            @Override
            public void onComplete() {
                onComplete.run();
            }
        };
    }

    public AtomicLong openR2dbcConnectionCounter = new AtomicLong();

    ConnectionFactory wrap(ConnectionFactory delegateCF) {
        return new ConnectionFactory() {
            @Override
            public Publisher<? extends io.r2dbc.spi.Connection> create() {
                return delegatePCS -> delegateCF.create().subscribe(subscriber(
                        delegatePCS::onSubscribe,
                        delegateC -> {
                            System.out.println("Open R2DBC connections (open) : " + openR2dbcConnectionCounter.incrementAndGet());
                            delegatePCS.onNext(new org.jooq.tools.r2dbc.LoggingConnection(new io.r2dbc.spi.Connection() {

                                // ---------------------------------------------------------------------
                                // 0.9.0.M1 API
                                // ---------------------------------------------------------------------

                                @Override
                                public Publisher<Void> beginTransaction() {
                                    return delegateC.beginTransaction();
                                }

                                @Override
                                public Publisher<Void> beginTransaction(TransactionDefinition definition) {
                                    return delegateC.beginTransaction(definition);
                                }

                                @Override
                                public Publisher<Void> close() {
                                    Publisher<Void> delegatePV = delegateC.close();
                                    return delegatePVS -> delegatePV.subscribe(subscriber(
                                            delegatePVS::onSubscribe,
                                            delegatePVS::onNext,
                                            t -> {
                                                System.out.println("Open R2DBC connections (error): " + openR2dbcConnectionCounter.decrementAndGet());
                                                t.printStackTrace();
                                                delegatePVS.onError(t);
                                            },
                                            () -> {
                                                System.out.println("Open R2DBC connections (close): " + openR2dbcConnectionCounter.decrementAndGet());
                                                delegatePVS.onComplete();
                                            }
                                    ));
                                }

                                @Override
                                public Publisher<Void> commitTransaction() {
                                    return delegateC.commitTransaction();
                                }

                                @Override
                                public Batch createBatch() {
                                    return delegateC.createBatch();
                                }

                                @Override
                                public Publisher<Void> createSavepoint(String name) {
                                    return delegateC.createSavepoint(name);
                                }

                                @Override
                                public io.r2dbc.spi.Statement createStatement(String sql) {
                                    return delegateC.createStatement(sql);
                                }

                                @Override
                                public boolean isAutoCommit() {
                                    return delegateC.isAutoCommit();
                                }

                                @Override
                                public ConnectionMetadata getMetadata() {
                                    return delegateC.getMetadata();
                                }

                                @Override
                                public IsolationLevel getTransactionIsolationLevel() {
                                    return delegateC.getTransactionIsolationLevel();
                                }

                                @Override
                                public Publisher<Void> releaseSavepoint(String name) {
                                    return delegateC.releaseSavepoint(name);
                                }

                                @Override
                                public Publisher<Void> rollbackTransaction() {
                                    return delegateC.rollbackTransaction();
                                }

                                @Override
                                public Publisher<Void> rollbackTransactionToSavepoint(String name) {
                                    return delegateC.rollbackTransactionToSavepoint(name);
                                }

                                @Override
                                public Publisher<Void> setAutoCommit(boolean autoCommit) {
                                    return delegateC.setAutoCommit(autoCommit);
                                }

                                @Override
                                public Publisher<Void> setTransactionIsolationLevel(IsolationLevel isolationLevel) {
                                    return delegateC.setTransactionIsolationLevel(isolationLevel);
                                }

                                @Override
                                public Publisher<Boolean> validate(ValidationDepth depth) {
                                    return delegateC.validate(depth);
                                }

                                // ---------------------------------------------------------------------
                                // 0.9.0.M2 API
                                // ---------------------------------------------------------------------

                                @Override
                                public Publisher<Void> setLockWaitTimeout(Duration timeout) {
                                    return delegateC.setLockWaitTimeout(timeout);
                                }

                                @Override
                                public Publisher<Void> setStatementTimeout(Duration timeout) {
                                    return delegateC.setStatementTimeout(timeout);
                                }
                            }));
                        },
                        delegatePCS::onError,
                        delegatePCS::onComplete
                ));
            }

            @Override
            public ConnectionFactoryMetadata getMetadata() {
                return delegateCF.getMetadata();
            }
        };
    }

    @ParameterizedTest
    @MethodSource("getCancellationDelays")
    void cancelShouldReleaseThePooledConnection(Duration cancellationDelay) throws InterruptedException {
        openR2dbcConnectionCounter.set(0);
        cancelQueryAfterDelay(sleepQuery(DSL.using(wrap(pooledConnectionFactory))), cancellationDelay);
        assertConnectionsReleased();
        assertEquals(0L, openR2dbcConnectionCounter.get());
    }

    @ParameterizedTest
    @MethodSource("getCancellationDelays")
    void workaroundShouldReleaseThePooledConnectionAfterCompletion(Duration cancellationDelay) throws InterruptedException {
        openR2dbcConnectionCounter.set(0);
        cancelQueryAfterDelay(notPropagatingCancellation(sleepQuery(DSL.using(wrap(pooledConnectionFactory)))), cancellationDelay);
        assertConnectionsReleased();
        assertEquals(0L, openR2dbcConnectionCounter.get());
    }

    private void assertConnectionsReleased() throws InterruptedException {
        // give plenty of time for `acquire` and `release` to occur
        Thread.sleep(1000);
        assertEquals(0, pooledConnectionFactory.getMetrics().get().acquiredSize());
    }

    @ParameterizedTest
    @MethodSource("getCancellationDelays")
    void cancelShouldCloseTheNonPooledConnection(Duration cancellationDelay) throws InterruptedException {
        openR2dbcConnectionCounter.set(0);
        ConnectionFactory connectionFactory = wrap(nonPooledConnectionFactory);
        cancelQueryAfterDelay(sleepQuery(DSL.using(connectionFactory)), cancellationDelay);
        Thread.sleep(1000L);
        assertEquals(0L, openR2dbcConnectionCounter.get());
    }

    @ParameterizedTest
    @MethodSource("getCancellationDelays")
    void workaroundShouldCloseTheNonPooledConnectionAfterCompletion(Duration cancellationDelay) throws InterruptedException {
        openR2dbcConnectionCounter.set(0);
        ConnectionFactory connectionFactory = wrap(nonPooledConnectionFactory);
        cancelQueryAfterDelay(notPropagatingCancellation(sleepQuery(DSL.using(connectionFactory))), cancellationDelay);
        Thread.sleep(100L);
        assertEquals(0L, openR2dbcConnectionCounter.get());
    }

    private static SelectSelectStep<Record1<Object>> sleepQuery(DSLContext dslContext) {
        return dslContext.select(function("pg_sleep", OTHER, value(SLEEP_SECONDS)));
    }

    private static void cancelQueryAfterDelay(Publisher<Record1<Object>> query, Duration cancellationDelay) {
        StepVerifier.create(query)
                .thenRequest(1)
                .thenAwait(cancellationDelay)
                .thenCancel()
                .verify();
    }

    private static Iterable<Duration> getCancellationDelays() {
        return List.of(Duration.ZERO,
                Duration.ofMillis(10),
                Duration.ofMillis(20),
                Duration.ofMillis(50),
                Duration.ofMillis(100),
                Duration.ofMillis(150)
        );
    }

    // the workaround
    static <T> Flux<T> notPropagatingCancellation(Publisher<T> query) {
        return Flux.create(sink -> Flux.from(query)
                .doOnNext(sink::next)
                .doOnError(sink::error)
                .doOnComplete(sink::complete)
                .subscribeOn(Schedulers.parallel())
                .subscribe());
    }
}
```

**Comment by lukaseder** [comment]
OK, I see, the problem is that the counter is decremented in the `onComplete` event of the `close()` publisher, not the `onNext` event. This produces a different result:

```java
                                @Override
                                public Publisher<Void> close() {
                                    Publisher<Void> delegatePV = delegateC.close();
                                    return delegatePVS -> delegatePV.subscribe(subscriber(
                                            delegatePVS::onSubscribe,
                                            t1 -> {
                                                System.out.println("Open R2DBC connections (next): " + openR2dbcConnectionCounter.decrementAndGet());
                                                delegatePVS.onNext(t1);
                                            },
                                            t -> {
                                                System.out.println("Open R2DBC connections (error): " + openR2dbcConnectionCounter.get());
                                                t.printStackTrace();
                                                delegatePVS.onError(t);
                                            },
                                            () -> {
                                                System.out.println("Open R2DBC connections (close): " + openR2dbcConnectionCounter.get());
                                                delegatePVS.onComplete();
                                            }
                                    ));
                                }
```

**Comment by lukaseder** [comment]
I've probably wasted way too much time with this issue, but I don't really see the bug here.
